modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/AppSchemaDataAccess.java

```diff
@@ -197,17 +198,23 @@ public FeatureTypeMapping getMappingByName(Name typeName) throws IOException {
      * @return
      * @throws IOException
      */
-    public FeatureTypeMapping getMappingByElement(Name typeName) throws IOException {
-        for (FeatureTypeMapping mapping : mappings.values()) {
-            if (mapping.getTargetFeature().getName().equals(typeName)) {
-                return mapping;
-            }
+    public FeatureTypeMapping getMappingByNameOrElement(Name typeName) throws IOException {
+        FeatureTypeMapping mapping = (FeatureTypeMapping) this.mappings.get(typeName);
+        if (mapping != null) {
+            return mapping;
         }
-        ArrayList<String> availables = new ArrayList<String>();
-        for (FeatureTypeMapping mapping : mappings.values()) {
-            availables.add(mapping.getTargetFeature().getName().toString());
+
+        // lookup by mapping name failed, try to lookup by target element
+        // NOTE: in this case, there is a risk of ambiguity. E.g. consider a (questionable) mapping configuration where multiple mappings of the
+        // same type have been specified, each with its own mappingName: they would have the same targetElement, and we wouldn't know which one to
+        // pick. The result would be unpredictable in such a situation.
```

Comment:
The problem I see in forbidding such a mapping configuration, is that it's neither wrong nor ambiguous, it's just more verbose.

The ambiguity is a consequence of using the element name to lookup the mapping, which in this case is not unique. I could throw an exception when the ambiguity cannot be resolved, but picking the first match has been the behavior  of the `getMappingByElement()` method so far and I'm afraid that changing it could break existing clients.